terralab/custom_http_server.py

```diff
@@ -0,0 +1,102 @@
+"""
+Custom local HTTP server logic used for OAuth2 flow
+"""
+
+from http.server import HTTPServer
+from typing import Optional
+from urllib.parse import parse_qs
+
+from oauth2_cli_auth.http_server import OAuthRedirectHandler
+from oauth2_cli_auth._timeout import TimeoutException, _method_with_timeout
+
+
+class CustomOAuthRedirectHandler(OAuthRedirectHandler):
+    """
+    HTTPRequest Handler that is intended to be used as oauth2 callback page.
+    Customized for terralab to receive and parse a POST request from the redirect.
+    """
+
+    def do_GET(self):
+        """For security reasons, we don't ever want to process a GET request."""
+        return NotImplementedError("GET request not supported for authentication flow.")
+
+    def do_POST(self) -> None:
+        ctype = self.headers.get("Content-Type")
+        if ctype == "application/x-www-form-urlencoded":
+            length = int(self.headers.get("Content-Length"))
+            raw_data = self.rfile.read(length)
+            params = parse_qs(raw_data.decode(encoding="utf_8"))
+        else:
+            params = {}
```

Comment:
should this throw an exception as well or log an error/warning?  seems like this would be a state we dont wanan be in

Reply:
this will end up giving them the "oh snap!" screen shown a few lines down, since "code" won't be in params, so has_error will be True. I think this is a good path.